internal/peering-request-operator/foreign-cluster.go

```diff
@@ -0,0 +1,152 @@
+package peering_request_operator
+
+import (
+	"errors"
+	v1 "github.com/liqoTech/liqo/api/discovery/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog"
+	"k8s.io/utils/pointer"
+	"strings"
+)
+
+func (r *PeeringRequestReconciler) UpdateForeignCluster(pr *v1.PeeringRequest) error {
+	tmp, err := r.crdClient.Resource("foreignclusters").Get(pr.Spec.ClusterID, metav1.GetOptions{})
+	if k8serrors.IsNotFound(err) {
+		// try to find it by label
+		err2 := err
+		tmp, err = r.crdClient.Resource("foreignclusters").List(metav1.ListOptions{
+			LabelSelector: "cluster-id=" + pr.Spec.ClusterID,
+		})
+		if err != nil {
+			klog.Error(err, err.Error())
+			return err
+		}
+		fcList, ok := tmp.(*v1.ForeignClusterList)
+		if !ok {
+			err = errors.New("retrieved object is not a ForeignClusterList")
+			klog.Error(err, err.Error())
+			return err
+		}
+		if len(fcList.Items) == 0 {
+			// object not found
+			err = err2
+		} else {
+			// objects retrieved
+			err = nil
+			tmp = &fcList.Items[0]
+		}
+	}
+
+	if k8serrors.IsNotFound(err) {
+		// create it
+		fc, err := r.createForeignCluster(pr)
+		if err != nil {
+			return err
+		}
+		r.setOwner(pr, fc)
+		return nil
+	} else if err != nil {
+		// other errors
+		klog.Error(err, err.Error())
+		return err
+	} else {
+		// update it
+		fc, ok := tmp.(*v1.ForeignCluster)
+		if !ok {
+			err = errors.New("retrieved object is not a ForeignCluster")
+			klog.Error(err, err.Error())
+			return err
+		}
+		if fc.Status.Incoming.PeeringRequest != nil {
+			// already up to date
+			return nil
+		}
+		fc.Status.Incoming.PeeringRequest = &corev1.ObjectReference{
+			Kind:       pr.Kind,
+			Name:       pr.Name,
+			UID:        pr.UID,
+			APIVersion: pr.APIVersion,
+		}
+		_, err = r.crdClient.Resource("foreignclusters").Update(fc.Name, fc, metav1.UpdateOptions{})
+		if err != nil {
+			klog.Error(err, err.Error())
+			return err
+		}
+		r.setOwner(pr, fc)
+		return nil
+	}
+}
+
+func (r *PeeringRequestReconciler) createForeignCluster(pr *v1.PeeringRequest) (*v1.ForeignCluster, error) {
+	var cnf *rest.Config
+	var err error
+
+	cnf, err = pr.GetConfig(r.crdClient.Client())
+	if err != nil {
+		if strings.Contains(err.Error(), "yaml") && r.Testing {
```

Comment:
this condition isn't robust at all; what if the error message changes?  Cannot we implement a more robust way to detect the error type?